fentik/flink
# Add support for stateless join on NULL key

Summary
---------

Issue: https://github.com/fentik/flink/issues/30

This pull request adds optimization for outer joins on a NULL key - while emitting the complete state, it evicts null records for these queries. As mentioned in https://github.com/fentik/flink/issues/30, Flink Join operators maintain
a complete set of all input data from left & right sides during a join. In the case of an outer equi join on a null key, we don't need to maintain as much state as the NULL value in the key will result in a failing join condition - in these cases, we can safely construct output rows without maintain any op state. This pull request builds this optimization for "Batch Backfill" mode - where when switching from batch to streaming (i.e. emitStateAndSwitchToStreaming), it cleans up state
for null keys -- thereby reducing operator state that is maintained. This PR also adds an execution option called
table.exec.stateless-join-on-null-key that allows you to control whether this optimization is enabled.

Test Plan
-------------
1) Execute following commands in Flink SQL: set table.exec.stateless-join-on-null-key='True'; set table.exec.batch-backfill='True';
2) Create NULL JOIN key in Postgress: update product set category_id=NULL where id=5699;
3) Execute LEFT OUTER JOIN in Flink SQL: select pc.id as category_id, p.id as product_id, product_name, category_name from product p left outer join product_category pc on p.category_id=pc.id;
4) Tail logs and make sure log line:
   "Performance Optimization - Evicting NULL record from state in outer
equijoin" is present
5) Run internal fentik regression test pipeline

See <img width="1767" alt="Screen Shot 2023-03-21 at 3 25 55 PM" src="https://user-images.githubusercontent.com/8899705/226766232-6a189ea6-b107-4f74-8f5c-05cac4576cf0.png">
where we see records get evicted from state
